### PR TITLE
Add Keccak variable length input spec

### LINKED ISSUE #143: Spec out Keccak Sponge part
This includes how Keccak takes multiple inputs and adds padding to them.

### PR DESCRIPTION
**Pull request by ChihChengLiang**
Fix #143

The lookup region is from @han0110's previous draft https://hackmd.io/oaZ_rzzET6GP1GG4Y8ojnQ?both

### CODE REVIEW

File: specs/keccak256/variable-length-input.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+#### Lookup
+
+To lookup the Keccak256 input to the output, query the following columns:
+
+- `is_end_result`: bool
+- `input`: RLC
+- `output`: RLC
+
+and constrain `is_end_result === 1`
```
Comment (ChihChengLiang) [inline specs/keccak256/variable-length-input.md:59]: @ed255 FYI this part might be the one you're looking for

File: specs/keccak256/variable-length-input.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
+# Keccak256 For Variable Length Input
+
+We describe the circuit that can process input of arbitrary size.
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:3]:
```suggestion
We describe the circuit that can process inputs of an arbitrary size.
```

File: specs/keccak256/variable-length-input.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+## Plain Keccak behavior
+
+Keccak256 use a simple iterating sponge construction to handle a variable length input. The sponge absorbs 17 words(136 bytes) of inputs, applies the `Keccak-f` permutation, then absorb next 17 words. Once all parts of the inputs are absorbed, the sponge squeezes out the 32 bytes for the output.
+
+If the last part of the inputs is less than 136 bytes, a padding must be applied. The padding is a multi-rate padding, it pads single bit 1 followed by minimum number of bits 0 followed by a single bit 1, such that the result is 136 bytes.
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:9]:
```suggestion
Keccak256 uses a simple iterating sponge construction to handle a variable length input. The sponge absorbs 17 words(136 bytes) of inputs, applies the `Keccak-f` permutation, and then absorb next 17 words. Once all parts of the inputs are absorbed, the sponge squeezes out the 32 bytes for the output.

If the last part of the input is less than 136 bytes, a padding must be applied. The padding is a multi-rate padding, it pads single bit 1 followed by minimum number of bits 0 followed by a single bit 1, such that the result is 136 bytes.
```

File: specs/keccak256/variable-length-input.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+- `output` The base-2 `state[:4]` output from this round `keccak_f`
+- `is_end_result` This flag indicates the hash output is final. The purpose is for external circuit to lookup.
+
+Selector:
+
+- `is_last_round_of_circuit` Last round of the whole circuit (not hash), to avoid `next` wrap around.
+
+#### Checks
+
+1. `hash_id` is sequential `next.id - curr.id in [0, 1]`
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:40]: I'm not sure if this is exactly needed here. This sounds more like a way to check that permutations are from the same hash operation. 

But that's already guaranteed by the accomulated length counters.

File: specs/keccak256/variable-length-input.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+2. If we are not in the last round of the circuit or last round of permutation
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:41]:
```suggestion
2. If we are not in the last round of the circuit
```

Not sure what the second part refers. But if it refers to the last round of the circuit too, I recommend to remove this as is more confusing.

File: specs/keccak256/variable-length-input.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+   1. Checks the `input_len` is  `curr.input_len === next.input_len`
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:42]:
```suggestion
   1. `curr.input_len === next.input_len`
```

File: specs/keccak256/variable-length-input.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+   2. `next.acc_len === curr.acc_len + 136`
+   3. `next.acc_len <= curr.input_len`
+   4. `next.acc_input === curr.acc_input * (r**136) + RLC(input, r)`
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:45]: I think this falls into the error of using RLC together with the witness within the same column. Is it possible? 
ping @han0110

File: specs/keccak256/variable-length-input.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+- `acc_len` How many length we have absorbed
+- `acc_input` Accumulatd bytes by random linear combination (in big-endian order)
+- `output` The base-2 `state[:4]` output from this round `keccak_f`
+- `is_end_result` This flag indicates the hash output is final. The purpose is for external circuit to lookup.
+
+Selector:
+
+- `is_last_round_of_circuit` Last round of the whole circuit (not hash), to avoid `next` wrap around.
+
+#### Checks
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:38]: I think we are missing a check for all of the rounds that will come after the padded one and which are essentially 0.

For all of these, we need to constrain the bytes to be 0 right? Or it does not affect since in the big circuit we will "dissable the permutation" when round > final_round_for_padding?

File: specs/keccak256/variable-length-input.md (hunk 5 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:41]: Also, would be nice to know if this refers to the last round that requires padding (last round with real witness) or simply the last round of the entire algorithm which might be full of 0's as the input might be much shorter and it is just used for padding constraints.

File: specs/keccak256/variable-length-input.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+- `is_end_result`: bool
+- `input`: RLC
+- `output`: RLC
+
+and constrain `is_end_result === 1`
+
+### Padding Region
+
+Note that we define a new `acc_len` which increament byte by byte, where the `acc_len` in the lookup region bumps by 136 bytes.
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:63]:
```suggestion
Note that we define a new `acc_len` which increments byte by byte, where the `acc_len` in the lookup region bumps by 136 bytes.
```

File: specs/keccak256/variable-length-input.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        return [0x81]
+    elif 1 < padlen < 136:
+        return [0x80] + ([0x00] * int(padlen - 2)) + [0x01]
+    else:
+        raise ValueError("unreachable")
+```
+
+#### Circuit
+
+The Padding Region is a 136-row region.
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:88]: I started implementing like 17 rows of 8bytes + 1 word for each. So 17 rows of 9 columns (with a few extra columns for the accomulators). Makes it much easier for the constraint appliance.

Can we copy into another region the required data to make the lookups? Or will increase the costs too much?

File: specs/keccak256/variable-length-input.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+- `byte` individual byte of the input in big-endian
+- `input_len` Length for correct padding
+- `acc_len` How many bytes we have processed
+- `condition_80_inv` The inverse of `input_len - acc_len` or 0.
+- `padded_byte` Mostly the same as the original `byte` but padded
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:94]: Not sure what `padded_byte` can be other than itself. You mean that it could be 0?

I say it because padded word would maybe make a bit more sense, but byte does not much to me.

**Review by CPerezz — COMMENTED** [review]
Just did a quick first round with some questions.

I'd appreciate also to have the table here or a link to it. Whatever you prefer. Would help to visualize what you did [here](https://hackmd.io/Ef0BFC4sQRGITNS-o5MiJA?both).

Finally, nice job!!! I think that this sums up very nicely all the discussions and ideas that have appeared and mixes them into the solution!

File: specs/keccak256/variable-length-input.md (hunk 11 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md:94]: For example, if the input length is 4 and it is `[0xff, 0xff, 0xff, 0xff]`. It would looks like

| byte | padded_byte |
| ---| --- |
|0xff|0xff|
|0xff|0xff|
|0xff|0xff|
|0xff|0xff|
|0x00|0x80|
|...|...|
|0x00|0x01|

**Comment by ChihChengLiang** [comment]
Added a table :)

File: specs/keccak256/variable-length-input.md (hunk 8 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md:38]: I think I have an idea for 1 now. Instead of complicating the design to handle unused permutations, the prover should add a dummy hash job `hash_id=3` the use it for 0 inputs for 97 permutations.

File: specs/keccak256/variable-length-input.md (hunk 4 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md]: We know this already due to the condition `input_len == acc_len` no?

We also know when to change when `acc_len_prev != acc_len_cur`

File: specs/keccak256/variable-length-input.md (hunk 7 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:45]: Maybe I'm mistaken as this is not used to perform an RLC later on.

IIRC, `  4. next.acc_input === curr.acc_input * (r**136) + RLC(input, r)` is checking that we're computing the RLC of the entire input correctly. 

if so, I think I was wrong with my concern and all is fine!

File: specs/keccak256/variable-length-input.md (hunk 8 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:38]: Yep, my concern was for 1. I'm not sure I follow you. Following @barryWhiteHat 's idea on that part, he purposed to do the following:

- If permutation contains initial input bytes -> `inp_len <= acc_len` for the last byte of the permutation:
1) ConditionallySelect(prev_output_state  || 0's) into actual state for the permutation computation
2) ConditionallySelect(actual_permutation_bytes || 0's) into actual permutation input
3) `final_result[0..N] += !is_last_permutation * out_state[0..N]
4) `running_sum_last_byte == output` NOTE that 3 and 4 are the same but with different formats.

I can elaborate it more. But will be something like saying, you conditionally copy to the next permutation as inputs `(state and next_inputs)` or a bunch of 0's. And you also conditionally add to `final_result` the result of the permutation depending on if it is the last one or not.

Point 2) sounds good as we discussed! :)

File: specs/keccak256/variable-length-input.md (hunk 10 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:92]: Ohhh. 

I started coding like:
- 17 rows of 9 columns
- Copy each byte to the padding_config region where we have the following columns:
- `byte`
- `total_len`
- `acc_len`
- `pad_len`
- `running_sum`

And is in this region where we will copy all bytes from offset 0-136 or even 0-MAX_INP_LEN as they have to be sequential for the entire hash.

WDYT?

Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:92]: The constraints are the same. The only thing that changes is how we layout the data. I did it like this as simplifies a lot more the things form the impl side. 

But maybe this is not inline with your proposal, and then I'm happy to refactor

File: specs/keccak256/variable-length-input.md (hunk 11 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:98]: So it's the permutation input what is padded. The byte can only be: 
- Itself
- 0
- 0x80
- 0x01
- (Special case) 0x81

File: specs/keccak256/variable-length-input.md (hunk 4 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md]: Not really, `acc_len` is bumped 136 each row. So it wouldn't exactly equal to `input_len`.

File: specs/keccak256/variable-length-input.md (hunk 11 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md:98]: Correct. I would consider 0 as the part of "Itself" though.

File: specs/keccak256/variable-length-input.md (hunk 10 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md:92]: Make sense to me. As long as the padding_config is 1 col 136 rows it should be fine. It's okay to copy `byte` col in padding_config to a 17-rows-9 columns gadget for other processes.

File: specs/keccak256/variable-length-input.md (hunk 4 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md]: Ohh, so `acc_len` is for permutation, not for all of the input bytes? I thought we had a variable that was couting the length for all of the bytes.
If it's not the case, yes, that's a way to do it with the id :)

File: specs/keccak256/variable-length-input.md (hunk 10 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:94]: Can't only be this. Needs some padding extra fields as I said.
Otherwise, you miss info to make the constraints between bytes.

Config looks like:
```rust
/// Constrains each byte of the keccak input to be padded correctly.
#[derive(Debug, Clone)]
pub(crate) struct PaddingBuilderConfig<F> {
    q_enable: Selector,
    // Contains:
    // - `byte` at Rotation(0)
    // - `pad_byte` at Rotation(1)
    byte: Column<Advice>,
    total_len: Column<Advice>,
    acc_len: Column<Advice>,
    // Contains:
    // - `pad_len` at Rotation(0)
    // - `pad_len_inv` at Rotation(1)
    pad_len: Column<Advice>,
    running_sum: Column<Advice>,
}
```
I can change this and avoid using rotations, having more columns and less rows.
Feel free to decide on this! :)

File: specs/keccak256/variable-length-input.md [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+   1. `next.input_len === curr.input_len`
+   2. `next.acc_len === curr.acc_len + 1`
+   3. Inverse check for `curr.condition_80_inv`
+   4. If `curr.input_len - curr.acc_len` is 0, pad `0x80`: `curr.padded_byte === curr.byte + (1 - (curr.input_len - curr.acc_len) * curr.condition_80_inv) * 0x80`
+4. For 1~135th rows
+   1. Set `is_pad_zone` to 1 if we entered. `curr.is_pad_zone === prev.is_pad_zone + (1 - (curr.input_len - curr.acc_len) * curr.condition_80_inv)`
+5. For the 135th row
+   1. Same as the previous 0x80 pad, but pad 0x01 if we are in the pad zone. `curr.padded_byte === curr.byte + (1 - (curr.input_len - curr.acc_len) * curr.condition_80_inv) * 0x80 + is_pad_zone * 0x01`
+6. Use another RLC gadget to check `byte` can be running summed up to `input` in the lookup region
+7. `padded_byte` are copied to a word builder gadget to build padded words, which would later be copied to the `Keccak-f` permutation
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:141]: This is a bit confusing. As some of the big points overlap the rows they touch. We could maybe group them.
Also, there's no mention to the special padding case with `0x81`

File: specs/keccak256/variable-length-input.md [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+The Padding Region is a 136-row region.
+
+- `byte` individual byte of the input in big-endian
+- `input_len` Length for correct padding
+- `acc_len` How many bytes we have processed
+- `condition_80_inv` The inverse of `input_len - acc_len` or 0.
+- `padded_byte` Mostly the same as the original `byte` but padded
+- `is_pad_zone` A flag to define the rows that `byte` should be 0
+
+| offset | byte | input_len | acc_len | condition_80_inv | padded_byte | is_pad_zone |
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:103]: Padding table is missing accumulator value which is the accomulative RLC.

**Review by CPerezz — COMMENTED** [review]
I think that would be really useful to:
1. Aside of having the entire circuit definition. Explain the components a bit more. Like sections for `padding`, `permutation`, `word_construction` etc...
2. Have a table showing the layout of each region mentioned above. On that way, is more similar to the specs we have for the other circuits.

WDYT?

File: specs/keccak256/variable-length-input.md (hunk 12 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md:141]: For big points 1~5, those checks have to be applied on their specific offsets. 6 is another gadget and 7 is all copy constraints.
I just realized point 4 can be grouped. Let me check.
The `0x81` case is in the 5.1. If it is a pad zone and condition_80 is activated then it would be `0x81`.

File: specs/keccak256/variable-length-input.md (hunk 10 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md:94]: Can we do "constraints between bytes" by copying them to 17 word-gadgets?

File: specs/keccak256/variable-length-input.md (hunk 13 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md:103]: Do we want to do them in this table or should we copy them to another RLC gadget as proposed in point 6? Because to accumulate RLC here we need to place the randomness in this column and complicate the logic.

File: specs/keccak256/variable-length-input.md (hunk 10 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:94]: I'm not sure I follow you here. 
The idea i was following (what I understood in our last call) was something along the lines of:

- Witness all of the bytes and padd correctly in `PermutationConfig`.
- From the previous config, we copy the info required to check that a byte is correctly padded into `PaddingBuilderConfig`.
- Finally, once all these constraints are applied, we copy each 8 bytes to the WordBuilder together with the word they form to constraint word construction correctness as well as that the bytes are indeed between [0-255].

Then we return the RLC accomulated for each byte of the permutation + the resulting State words.

File: specs/keccak256/variable-length-input.md (hunk 12 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:141]: Makes sense! Thanks!!!

File: specs/keccak256/variable-length-input.md (hunk 13 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:103]: We can have the randomness in an instance column as @han0110 suggested. 
The only issue I see is that we have a lot of rows, which will mean that we have A LOT of instance rows too. And generally we want to have as less PI as possible.

Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:103]: We can otherways just copy the randomness once, and copy_constraint it from row to row. So that we make sure that the randomness used is always the same queried from the initial instance column.

File: specs/keccak256/variable-length-input.md (hunk 12 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md]: Addressed in [40e0186](https://github.com/appliedzkp/zkevm-specs/pull/147/commits/40e0186951315ab16eca26dfcb932f0b15b19933)

File: specs/keccak256/variable-length-input.md (hunk 10 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md:100]: Resolved offline. We agreed to go naive and defer the optimization later.
The optimization is about merging `byte` and `pad_byte` and using `Rotation(2)` to reduce columns and with the cost of increasing rotation, which is more cost-efficient in the aggregation circuit. But merging the columns would make debugging gates harder and optimization at this stage is a bit early.

File: specs/keccak256/variable-length-input.md (hunk 13 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md]: Added the RLC column for `byte` in [87f13e6](https://github.com/appliedzkp/zkevm-specs/pull/147/commits/87f13e620cd9c159e2d031bba1d844d88c777920)

File: specs/keccak256/variable-length-input.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+Define:
+
+We use `Round` for the cells of the following columns in a row, it represents the information in a `keccak_f` round.
+
+The `curr: Round` and `next: Round` represent the current and the next row.
+
+Columns:
+
+- `hash_id` Sequential number for grouping different hash
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:26]: I'm still not sure how can we have two `hash_id`s within the same circuit instance. In case we have it because all the permutations of all hashes will share some common region/table, couldn't we use the RLC(input) as id? I think makes a bit more sense as it ties the input to the permutation computations.

File: specs/keccak256/variable-length-input.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+- `is_end_result` This flag indicates the hash output is final. The purpose is for external circuit to lookup.
+
+Selector:
+
+- `is_last_round_of_circuit` Last round of the whole circuit (not hash), to avoid `next` wrap around.
+
+| hash_id | input_len | input | acc_len | acc_input | output | is_end_result | output_word_0 | output_word_1 | output_word_2 | output_word\_... |
+| ------: | --------: | ----- | ------: | --------- | ------ | ------------: | ------------- | ------------- | ------------- | ---------------- |
+|       0 |       150 |       |       0 |           |        |             0 |               |               |               |                  |
+|       0 |       150 |       |     136 |           |        |             1 |               |               |               |                  |
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:42]: I'm not following this table at all. Is this related to an entire hash?? And here we have all the permutations of all of the hashes? If that's the case, Is this trying to describe the FULL keccak hash? Including N permutations per call?

File: specs/keccak256/variable-length-input.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+The `curr: Round` and `next: Round` represent the current and the next row.
+
+Columns:
+
+- `hash_id` Sequential number for grouping different hash
+- `input_len` Length for correct padding
+- `input` 136 bytes to be absorbed in this round
+- `acc_len` How many length we have absorbed
+- `acc_input` Accumulatd bytes by random linear combination (in big-endian order)
```
Comment (CPerezz) [inline specs/keccak256/variable-length-input.md:30]: In this context, aren't these the same? We have absorbed as many bytes as we have accomulated in the RLC no?

File: specs/keccak256/variable-length-input.md (hunk 15 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:42]: If I'm correct, then `input` can't be a single column but rather 17 word columns or something similar no?

**Review by CPerezz — COMMENTED** [review]
Some concerns about the general table that I'm not sure I do follow. 

Sorry for that many iterations @ChihChengLiang :pray:

**Comment by ChihChengLiang** [comment]
Per discussion with @ed255 in https://github.com/appliedzkp/zkevm-specs/pull/150#discussion_r829974017, we need `is_enabled` column.

File: specs/keccak256/variable-length-input.md (hunk 14 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md:27]: Using input as the id might not make sense.
The definition of the input is "the RLC of the 136 bytes to be absorbed in this round". Let's say if we have two rows with the same input value 123,  their output could be different if one is the first 136 bytes of the acc input and the other is the second 136 bytes of the acc input.

File: specs/keccak256/variable-length-input.md (hunk 15 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md]: Sorry, can you point to the columns of the cells again? I'm not following.

File: specs/keccak256/variable-length-input.md (hunk 16 above)
Comment (ChihChengLiang) [reply specs/keccak256/variable-length-input.md:31]: The `acc_len` is the length of bytes, while `acc_input` is RLC. `acc_len` can be 0 or a multiple of 136, we use it to track if a hash input ends at `input_len`.

**Comment by ChihChengLiang** [comment]
Added `input_len` to the lookup columns, as discussed with @han0110 in https://github.com/appliedzkp/zkevm-specs/pull/150#discussion_r830135134.

File: specs/keccak256/variable-length-input.md (hunk 14 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:27]: But within the same permutation input computation, we already enforce the continuity and the relation between each byte by constraining the RLC.

If the RLC matches, the bytes were for sure from the same ID no?

File: specs/keccak256/variable-length-input.md (hunk 16 above)
Comment (CPerezz) [reply specs/keccak256/variable-length-input.md:31]: ˋacc_inputˋ is always [0,135]??

**Review by CPerezz — APPROVED** [review]
Just final questions over `acc_len` and the `hash_id`. Once resolved, feel free to merge!

**Comment by ChihChengLiang** [comment]
close in favor of #161
